File: IOI/IOI 13-Wombats/grader.c
#include <stdio.h>
#include <stdlib.h>
#include "wombats.h"

#define fail(s, x...) do { \
		fprintf(stderr, s "\n", ## x); \
		exit(1); \
	} while(0)

static int H[5000][200];
static int V[5000][200];

int main() {
	int R, C, E, P, Q, W, V1, V2, event, i, j;
	int res;

	FILE *f = fopen("wombats.in", "r");
	if (!f)
		fail("Failed to open input file.");

	res = fscanf(f, "%d%d", &R, &C);
    for (i = 0; i < R; ++i)
        for (j = 0; j < C-1; ++j)
            res = fscanf(f, "%d", &H[i][j]);
    for (i = 0; i < R-1; ++i)
        for (j = 0; j < C; ++j)
            res = fscanf(f, "%d", &V[i][j]);

    init(R, C, H, V);

	res = fscanf(f, "%d", &E);
	for (i = 0; i < E; i++) {
		res = fscanf(f, "%d", &event);
        if (event == 1) {
            res = fscanf(f, "%d%d%d", &P, &Q, &W);
            changeH(P, Q, W);
        } else if (event == 2) {
            res = fscanf(f, "%d%d%d", &P, &Q, &W);
            changeV(P, Q, W);
        } else if (event == 3) {
            res = fscanf(f, "%d%d", &V1, &V2);
            printf("%d\n", escape(V1, V2));
        } else
            fail("Invalid event type.");
	}

	fclose(f);
	return 0;
}

File: IOI/IOI 13-Wombats/grader.pas
program grader;

uses wombats;

procedure fail(s: string);
begin
    writeln(stderr, s);
    halt(1);
end;

var
    R, C, E, P, Q, W, V1, V2, event, i, j: longint;
    H, V: array[0..4999, 0..199] of LongInt;
    f: text;

begin
    assign(f, 'wombats.in');
    {$I-} reset(f); {$I+}
    if IOResult <> 0 then
        fail('Failed to open input file.');

    readln(f, R, C);
    for i := 0 to R-1 do
        for j := 0 to C-2 do
            read(f, H[i, j]);
    for i := 0 to R-2 do
        for j := 0 to C-1 do
            read(f, V[i, j]);

    init(R, C, H, V);

    read(f, E);
    for i := 1 to E do begin
        read(f, event);
        if event = 1 then begin
            read(f, P, Q, W);
            changeH(P, Q, W);
        end else if event = 2 then begin
            read(f, P, Q, W);
            changeV(P, Q, W);
        end else if event = 3 then begin
            read(f, V1, V2);
            writeln(escape(V1, V2));
        end else
            fail('Invalid event type.')
    end;

    close(f);
end.
